### PR TITLE
[Flang][OpenMP][MLIR][Lower] Update lowering to use wrapper ops

### PR DESCRIPTION
**Pull request by skatrak**
This patch introduces the following changes:
  - Complete OpenMP directive sets to include masked and taskloop-related ones.
  - Refactor clause processing to store related operands in structures. These are combined into construct-specific operand structures using a mixin pattern and simplify keeping track of operand lists used for creating OpenMP operations as well as reducing argument lists.
  - Update the lowering to MLIR of loop constructs to follow the wrapper + canonical loop approach. This includes some changes to early privatization (as opposed to "delayed privatization") that need testing.
  - Add lowering for the simplified omp.canonical_loop operation, based on omp.wsloop.
  - Significant refactor of OpenMP construct lowering, separating composite constructs into their own "gen" functions and handling combined loop constructs in a more scalable way. Updated genOMP functions to follow the same set of patterns to make it more understandable. Split clause processing calls from code generation for operations in preparation for reusing the same code when dealing with composite constructs.
  - Add basic support for taskloop, for completeness sake.
  - Document missing clauses.
  - Move some code from OpenMP.cpp to ClauseProcessor.cpp and call it there to simplify calls to certain `process...()` methods that were always followed by the same post-processing.
  - Add and populate "composite" attribute on wrapper operations.
  - Add empty constructors for wrapper operations.
  - Update `LoopWrapperInterface` to avoid compilation problems.
  - Update `Task*` operation's names to follow the same capitalization convention.
  - Changes to SCF to OpenMP conversion to produce a wrapper omp.wsloop with a nested omp.canonical_loop rather than an invalid omp.wsloop.
  - Document missing verifier checks on wrapper operations.

All these changes are still WIP, as they haven't been tested. It's also a TODO to support composite constructs, though almost everything is in place in order to allow that. I'll update this PR after I spend some time testing it and add support for composite constructs.

### CODE REVIEW

File: flang/include/flang/Semantics/openmp-directive-sets.h [hunk 1]
```diff
@@ -24,68 +24,80 @@ namespace llvm::omp {
...
-    Directive::OMPD_parallel_do,
     Directive::OMPD_parallel_do_simd,
+    Directive::OMPD_parallel_do,
+    Directive::OMPD_parallel_masked_taskloop_simd,
+    Directive::OMPD_parallel_masked_taskloop,
+    Directive::OMPD_parallel_master_taskloop_simd,
+    Directive::OMPD_parallel_master_taskloop,
     Directive::OMPD_parallel_sections,
     Directive::OMPD_parallel_workshare,
+    Directive::OMPD_parallel,
```
Comment (kparzysz) [inline flang/include/flang/Semantics/openmp-directive-sets.h:35]: How are you sorting these directives?  When I run "sort" (the Unix command), I get the OMPD_parallel first in the output.

File: flang/include/flang/Semantics/openmp-directive-sets.h
Comment (kparzysz) [inline flang/include/flang/Semantics/openmp-directive-sets.h:1]: This change may be a standalone PR.

File: flang/lib/Lower/OpenMP/ReductionProcessor.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ ReductionProcessor::ReductionIdentifier ReductionProcessor::getReductionType(
 
 void ReductionProcessor::addReductionSym(
     const Fortran::parser::OmpReductionClause &reduction,
-    llvm::SmallVector<const Fortran::semantics::Symbol *> &symbols) {
+    llvm::SmallVectorImpl<const Fortran::semantics::Symbol *> &symbols) {
```
Comment (kparzysz) [inline flang/lib/Lower/OpenMP/ReductionProcessor.cpp:63]: The SmallVector -> SmallVectorImpl here and below could be a separate PR.

File: flang/lib/Lower/OpenMP/ReductionProcessor.h [hunk 3]
```diff
@@ -101,7 +101,7 @@ class ReductionProcessor {
 
   static void addReductionSym(
       const Fortran::parser::OmpReductionClause &reduction,
-      llvm::SmallVector<const Fortran::semantics::Symbol *> &symbols);
+      llvm::SmallVectorImpl<const Fortran::semantics::Symbol *> &symbols);
```
Comment (kparzysz) [inline flang/lib/Lower/OpenMP/ReductionProcessor.h:104]: The other SmallVector -> SmallVectorImpl change.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 4]
```diff
@@ -1122,7 +1211,7 @@ def TaskGroupOp : OpenMP_Op<"taskgroup", [AttrSizedOperandSegments,
 // 2.10.4 taskyield Construct
 //===----------------------------------------------------------------------===//
 
-def TaskyieldOp : OpenMP_Op<"taskyield"> {
+def TaskYieldOp : OpenMP_Op<"taskyield"> {
```
Comment (kparzysz) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1214]: This and the TaskWait rename could be a separate PR.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 5]
```diff
@@ -1337,10 +1426,10 @@ def MapInfoOp : OpenMP_Op<"map_info", [AttrSizedOperandSegments]> {
     - `var_type`: The type of the variable to copy.
     - `var_ptr_ptr`: Used when the variable copied is a member of a class, structure
       or derived type and refers to the originating struct.
-    - `members`:  Used to indicate mapped child members for the current MapInfoOp,
```
Comment (kparzysz) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1340]: Trailing whitespace removal could be a quick PR on its own.

**Review by kparzysz — COMMENTED** [review]
Could you isolate the addition of CanonicalLoop op into a separate PR?

Overall the changes look ok, but there are several logical groups of changes that make more sense in separate PRs.

File: flang/include/flang/Semantics/openmp-directive-sets.h (hunk 1 above)
Comment (skatrak) [reply flang/include/flang/Semantics/openmp-directive-sets.h:35]: I used VSCode's sort lines command, but I'm not quite sure what happened. It might have been related to the commas, I think. I tried again without them and it looks like the ordering made more sense. Just created an upstream PR with these changes.

**Comment by skatrak** [comment]
Features in this PR already landed via merge from main.
